cdk/src/open-data-platform/data-plane/data-import/write-violations-data.handler.ts

```diff
@@ -0,0 +1,260 @@
+): Promise<number> {
+  return new Promise(function (resolve, reject) {
+    const batchSize = 10;
+    let numberRows = 0;
+    let results: ViolationsTableRow[] = [];
+
+    const fileStream = S3.getObject(s3Params).createReadStream();
+    let pipeline = chain([
+      fileStream,
+      Pick.withParser({ filter: 'features' }),
+      streamArray(),
+      new Batch({ batchSize: batchSize }),
+    ]);
+
+    pipeline
+      .on('data', async (batch: any[]) => {
+        if (numberRows < numberOfRowsToWrite) {
+          for (const data of batch) {
+            const value = data.value;
+            const properties = value.properties;
+
+            console.log(properties);
+            if (LEAD_AND_COPPER_VIOLATIONS.has(properties[VIOLATION_CODE])) {
+              const startDate = moment(properties[COMPL_PER_BEGIN_DATE], 'DD-MMM-YY');
+              const row = new ViolationsTableRowBuilder()
+                .violationId(properties[VIOLATION_ID])
+                .pwsId(properties[PWSID])
+                .violationCode(properties[VIOLATION_CODE])
+                .complianceStatus(
+                  COMPLIANCE_STATUS_MAP.get(properties[COMPLIANCE_STATUS_CODE]) ?? '',
+                )
+                .startDate(startDate.format('YYYY-MM-DD'))
+                .geom(JSON.stringify(value.geometry))
+                .build();
+              results.push(row);
+
+              // Every batch size, write into the db.
+              if (results.length == batchSize) {
+                // Pause reads while inserting into db.
+                fileStream.pause();
+                await insertRows(db, results);
+                results = [];
+                fileStream.resume();
+              }
+            }
+          }
+        } else {
+          // Stop reading stream if numberOfRowsToWrite has been met.
+          fileStream.destroy();
+        }
+        numberRows += batch.length;
+      })
+      .on('error', (error: Error) => {
+        reject(error);
+      })
+      // Gets called by pipeline.destroy()
+      .on('close', async (_: Error) => {
+        resolve(numberRows);
+      })
+      .on('end', async () => {
+        resolve(numberRows);
+      });
+  });
+}
+
+/**
+ * Parses S3 'violations_by_water_system.csv' file and writes rows
+ * to violations table in the MainCluster postgres db.
+ */
+export async function handler(_: APIGatewayProxyEvent): Promise<APIGatewayProxyResult> {
+  const numberRowsToWrite: number = parseInt(
+    process.env.numberRows ?? `${DEFAULT_NUMBER_ROWS_TO_INSERT}`,
+  );
+
+  const s3Params = {
+    Bucket: 'opendataplatformapistaticdata',
+    Key: 'violations_with_geom.geojson',
+  };
+
+  let db: ConnectionPool | undefined;
+
+  // Read CSV file and write to violations table.
+  try {
+    db = await connectToDb();
+    if (db == undefined) {
+      throw Error('Unable to connect to db');
+    }
+
+    // Remove existing rows before inserting new ones.
+    await deleteRows(db);
```

Comment:
(out of curiosity) why do we need to delete? couldn't we just upsert rows?

Reply:
Good question - switched to upsert